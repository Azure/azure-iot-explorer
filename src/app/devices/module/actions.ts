/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import actionCreatorFactory from 'typescript-fsa';
import * as actionPrefixes from '../../constants/actionPrefixes';
import * as actionTypes from '../../constants/actionTypes';
import { ModuleIdentity } from './../../api/models/moduleIdentity';
import { ModuleTwin } from './../../api/models/moduleTwin';

const moduleIdentityCreator = actionCreatorFactory(actionPrefixes.MODULEIDENTITY);
const getModuleIdentitiesAction = moduleIdentityCreator.async<string, ModuleIdentity[]>(actionTypes.GET_MODULE_IDENTITIES);
const addModuleIdentityAction = moduleIdentityCreator.async<ModuleIdentity, ModuleIdentity>(actionTypes.ADD_MODULE_IDENTITY);
const getModuleIdentityTwinAction = moduleIdentityCreator.async<GetModuleIdentityTwinActionParameters, ModuleTwin>(actionTypes.GET_MODULE_IDENTITY_TWIN);
const getModuleIdentityAction = moduleIdentityCreator.async<GetModuleIdentityActionParameters, ModuleIdentity>(actionTypes.GET_MODULE_IDENTITY);
const deleteModuleIdentityAction = moduleIdentityCreator.async<DeleteModuleIdentityActionParameters, void>(actionTypes.DELETE_MODULE_IDENTITY);

export {
    addModuleIdentityAction,
    deleteModuleIdentityAction,
    getModuleIdentitiesAction,
    getModuleIdentityAction,
    getModuleIdentityTwinAction
};

export interface GetModuleIdentityTwinActionParameters {
    moduleId: string;
    deviceId: string;
}

export interface GetModuleIdentityActionParameters {
    moduleId: string;
    deviceId: string;
}

export interface DeleteModuleIdentityActionParameters {
    moduleId: string;
    deviceId: string;
}
