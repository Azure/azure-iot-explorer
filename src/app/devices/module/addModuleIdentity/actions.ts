/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import actionCreatorFactory from 'typescript-fsa';
import { MODULEIDENTITY } from '../../../constants/actionPrefixes';
import { ADD_MODULE_IDENTITY } from '../../../constants/actionTypes';
import { ModuleIdentity } from '../../../api/models/moduleIdentity';

export const addModuleIdentityAction = actionCreatorFactory(MODULEIDENTITY).async<ModuleIdentity, ModuleIdentity>(ADD_MODULE_IDENTITY);
