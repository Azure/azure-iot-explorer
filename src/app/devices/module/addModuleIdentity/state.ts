/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { Record } from 'immutable';
import { SynchronizationStatus } from '../../../api/models/synchronizationStatus';
import { AddDeviceStateInterface } from '../../addDevice/state';
import { IM } from '../../../shared/types/types';

export interface AddModuleStateInterface {
    synchronizationStatus: SynchronizationStatus;
}

export const addModuleStateInitial = Record<AddDeviceStateInterface>({
    synchronizationStatus: SynchronizationStatus.initialized
});

export type AddModuleStateType = IM<AddModuleStateInterface>;
