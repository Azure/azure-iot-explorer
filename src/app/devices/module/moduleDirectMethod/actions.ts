/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import actionCreatorFactory from 'typescript-fsa';
import { DEVICECONTENT } from '../../../constants/actionPrefixes';
import { INVOKE_MODULE_IDENTITY_METHOD } from '../../../constants/actionTypes';

export const invokeModuleDirectMethodAction = actionCreatorFactory(DEVICECONTENT).async<InvokeModuleMethodActionParameters, string>(INVOKE_MODULE_IDENTITY_METHOD);

export interface InvokeModuleMethodActionParameters {
    connectTimeoutInSeconds: number;
    deviceId: string;
    methodName: string;
    moduleId: string;
    payload?: any; // tslint:disable-line:no-any
    responseTimeoutInSeconds: number;
}
