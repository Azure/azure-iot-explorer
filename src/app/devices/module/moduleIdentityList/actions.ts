/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import actionCreatorFactory from 'typescript-fsa';
import { MODULEIDENTITY } from '../../../constants/actionPrefixes';
import { GET_MODULE_IDENTITIES } from '../../../constants/actionTypes';
import { ModuleIdentity } from '../../../api/models/moduleIdentity';

export const getModuleIdentitiesAction = actionCreatorFactory(MODULEIDENTITY).async<string, ModuleIdentity[]>(GET_MODULE_IDENTITIES);
