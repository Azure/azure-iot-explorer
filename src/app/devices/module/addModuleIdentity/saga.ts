/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { call, put, takeEvery } from 'redux-saga/effects';
import { SagaIterator } from 'redux-saga';
import { Action } from 'typescript-fsa';
import { addModuleIdentity } from '../../../api/services/moduleService';
import { raiseNotificationToast } from '../../../notifications/components/notificationToast';
import { NotificationType } from '../../../api/models/notification';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import { addModuleIdentityAction } from './actions';
import { ModuleIdentity } from './../../../api/models/moduleIdentity';

export function* addModuleIdentitySagaWorker(action: Action<ModuleIdentity>): SagaIterator {
    try {
        const parameters = {
            moduleIdentity: action.payload,
        };

        const moduleIdentity = yield call(addModuleIdentity, parameters);

        yield call(raiseNotificationToast, {
            text: {
                translationKey: ResourceKeys.notifications.addModuleIdentityOnSucceed,
                translationOptions: {
                    moduleId: action.payload.moduleId
                },
            },
            type: NotificationType.success
        });

        yield put(addModuleIdentityAction.done({params: action.payload, result: moduleIdentity}));
    } catch (error) {
        yield call(raiseNotificationToast, {
            text: {
                translationKey: ResourceKeys.notifications.addModuleIdentityOnError,
                translationOptions: {
                    error,
                    moduleId: action.payload.moduleId
                },
            },
            type: NotificationType.error
          });

        yield put(addModuleIdentityAction.failed({params: action.payload, error}));
    }
}

export function* addModuleIdentitySaga() {
    yield takeEvery(addModuleIdentityAction.started.type, addModuleIdentitySagaWorker);
}
