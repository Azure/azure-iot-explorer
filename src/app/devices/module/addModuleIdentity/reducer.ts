/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { reducerWithInitialState } from 'typescript-fsa-reducers';
import { addModuleStateInitial, AddModuleStateType } from './state';
import { addModuleIdentityAction } from './actions';
import { SynchronizationStatus } from '../../../api/models/synchronizationStatus';

export const addModuleIdentityReducer = reducerWithInitialState<AddModuleStateType>(addModuleStateInitial())
    .case(addModuleIdentityAction.started, (state: AddModuleStateType) => {
        return state.merge({
            synchronizationStatus: SynchronizationStatus.working
        });
    })
    .case(addModuleIdentityAction.done, (state: AddModuleStateType) => {
        return state.merge({
            synchronizationStatus: SynchronizationStatus.upserted
        });
    })
    .case(addModuleIdentityAction.failed, (state: AddModuleStateType) => {
        return state.merge({
            synchronizationStatus: SynchronizationStatus.failed
        });
    });
