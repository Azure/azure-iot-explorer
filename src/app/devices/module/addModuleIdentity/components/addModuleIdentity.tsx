/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { useLocation, useHistory } from 'react-router-dom';
import { CommandBar, ChoiceGroup, IChoiceGroupOption, Checkbox, TextField } from '@fluentui/react';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { getDeviceIdFromQueryString } from '../../../../shared/utils/queryStringHelper';
import { CANCEL, SAVE } from '../../../../constants/iconNames';
import { SynchronizationStatus } from '../../../../api/models/synchronizationStatus';
import { ROUTE_PARAMS } from '../../../../constants/routes';
import { HeaderView } from '../../../../shared/components/headerView';
import { DeviceAuthenticationType } from '../../../../api/models/deviceAuthenticationType';
import { validateThumbprint, validateKey, validateModuleIdentityName, processThumbprint } from '../../../../shared/utils/utils';
import { useAsyncSagaReducer } from '../../../../shared/hooks/useAsyncSagaReducer';
import { addModuleIdentityReducer } from '../reducer';
import { addModuleIdentitySaga } from '../saga';
import { addModuleStateInitial } from '../state';
import { addModuleIdentityAction } from '../actions';
import { AppInsightsClient } from '../../../../shared/appTelemetry/appInsightsClient';
import { TELEMETRY_USER_ACTIONS } from '../../../../constants/telemetry';
import '../../../../css/_deviceDetail.scss';

const initialKeyValue = {
    error: '',
    thumbprint: '',
    thumbprintError: '',
    value: ''
};

export const AddModuleIdentity: React.FC = () => {
    const { t } = useTranslation();
    const { search, pathname } = useLocation();
    const history = useHistory();
    const deviceId = getDeviceIdFromQueryString(search);

    const [ localState, dispatch ] = useAsyncSagaReducer(addModuleIdentityReducer, addModuleIdentitySaga, addModuleStateInitial(), 'addModuleState');
    const { synchronizationStatus } = localState;
    const [ authenticationType, setAuthenticationType ] = React.useState(DeviceAuthenticationType.SymmetricKey);
    const [ autoGenerateKeys, setautoGenerateKeys ] = React.useState<boolean>(true);
    const [ module, setModule ] = React.useState<{ id: string, error: string }>({ id: '', error: ''});
    const [ primaryKey, setPrimaryKey ] = React.useState(initialKeyValue);
    const [ secondaryKey, setSecondaryKey ] = React.useState(initialKeyValue);

    React.useEffect(() => {
        if (synchronizationStatus === SynchronizationStatus.upserted) { // only when module identity has been added successfully would navigate to module list view
            navigateToModuleList();
        }
    },              [synchronizationStatus]);

    const showCommandBar = () => {
        return (
            <CommandBar
                className="command"
                items={[
                    {
                        ariaLabel: t(ResourceKeys.moduleIdentity.command.save),
                        disabled: disableSaveButton(),
                        iconProps: {iconName: SAVE},
                        key: SAVE,
                        name: t(ResourceKeys.moduleIdentity.command.save),
                        onClick: handleSave
                    },
                    {
                        ariaLabel: t(ResourceKeys.moduleIdentity.command.cancel),
                        iconProps: {iconName: CANCEL},
                        key: CANCEL,
                        name: t(ResourceKeys.moduleIdentity.command.cancel),
                        onClick: navigateToModuleList
                    }
                ]}
            />
        );
    };

    const showModuleId = () => {
        return (
            <TextField
                ariaLabel={t(ResourceKeys.moduleIdentity.moduleId)}
                label={t(ResourceKeys.moduleIdentity.moduleId)}
                value={module.id}
                required={true}
                onChange={changeModuleIdentityName}
                errorMessage={!!module.error ? t(module.error) : ''}
                description={t(ResourceKeys.moduleIdentity.moduleIdTooltip)}
            />
        );
    };

    const getAuthType = () => {
        return (
            <ChoiceGroup
                label={t(ResourceKeys.moduleIdentity.authenticationType.text)}
                selectedKey={authenticationType}
                onChange={changeAuthenticationType}
                options={
                    [
                        {
                            key: DeviceAuthenticationType.SymmetricKey,
                            text: t(ResourceKeys.moduleIdentity.authenticationType.symmetricKey.type)
                        },
                        {
                            key: DeviceAuthenticationType.SelfSigned,
                            text: t(ResourceKeys.moduleIdentity.authenticationType.selfSigned.type)
                        },
                        {
                            key: DeviceAuthenticationType.CACertificate,
                            text: t(ResourceKeys.moduleIdentity.authenticationType.ca.type)
                        }
                    ]
                }
                required={true}
            />
        );
    };

    const renderSymmetricKeySection = () => {
        return (
            <>
                <TextField
                    ariaLabel={t(ResourceKeys.moduleIdentity.authenticationType.symmetricKey.primaryKey)}
                    label={t(ResourceKeys.moduleIdentity.authenticationType.symmetricKey.primaryKey)}
                    value={primaryKey.value}
                    required={true}
                    onChange={changePrimaryKey}
                    type={'password'}
                    canRevealPassword={true}
                    revealPasswordAriaLabel={t(ResourceKeys.common.revealPassword)}
                    errorMessage={!!primaryKey.error ? t(primaryKey.error) : ''}
                    description={t(ResourceKeys.moduleIdentity.authenticationType.symmetricKey.primaryKeyTooltip)}
                />
                <TextField
                    ariaLabel={t(ResourceKeys.moduleIdentity.authenticationType.symmetricKey.secondaryKey)}
                    label={t(ResourceKeys.moduleIdentity.authenticationType.symmetricKey.secondaryKey)}
                    value={secondaryKey.value}
                    required={true}
                    onChange={changeSecondaryKey}
                    type={'password'}
                    canRevealPassword={true}
                    revealPasswordAriaLabel={t(ResourceKeys.common.revealPassword)}
                    errorMessage={!!secondaryKey.error ? t(secondaryKey.error) : ''}
                    description={t(ResourceKeys.moduleIdentity.authenticationType.symmetricKey.secondaryKeyTooltip)}
                />
            </>
        );
    };

    const renderSelfSignedSection = () => {
        return (
            <>
                <TextField
                    ariaLabel={t(ResourceKeys.moduleIdentity.authenticationType.selfSigned.primaryThumbprint)}
                    label={t(ResourceKeys.moduleIdentity.authenticationType.selfSigned.primaryThumbprint)}
                    value={primaryKey.thumbprint}
                    required={true}
                    onChange={changePrimaryThumbprint}
                    type={'password'}
                    canRevealPassword={true}
                    revealPasswordAriaLabel={t(ResourceKeys.common.revealPassword)}
                    errorMessage={!!primaryKey.thumbprintError ? t(primaryKey.thumbprintError) : ''}
                    description={t(ResourceKeys.moduleIdentity.authenticationType.selfSigned.primaryThumbprintTooltip)}
                />
                <TextField
                    ariaLabel={t(ResourceKeys.moduleIdentity.authenticationType.selfSigned.secondaryThumbprint)}
                    label={t(ResourceKeys.moduleIdentity.authenticationType.selfSigned.secondaryThumbprint)}
                    value={secondaryKey.thumbprint}
                    required={true}
                    onChange={changeSecondaryThumbprint}
                    type={'password'}
                    canRevealPassword={true}
                    revealPasswordAriaLabel={t(ResourceKeys.common.revealPassword)}
                    readOnly={false}
                    errorMessage={!!secondaryKey.thumbprintError ? t(secondaryKey.thumbprintError) : ''}
                    description={t(ResourceKeys.moduleIdentity.authenticationType.selfSigned.secondaryThumbprintTooltip)}
                />
            </>
        );
    };

    const showAuthentication = () => {
        return (
            <div className="authentication">
                {getAuthType()}
                {authenticationType === DeviceAuthenticationType.SymmetricKey &&
                    <>
                        <Checkbox
                            className="autoGenerateButton"
                            label={t(ResourceKeys.moduleIdentity.authenticationType.symmetricKey.autoGenerate)}
                            checked={autoGenerateKeys}
                            onChange={changeAutoGenerateKeys}
                        />
                        {!autoGenerateKeys &&
                            renderSymmetricKeySection()
                        }
                    </>
                }
                {authenticationType === DeviceAuthenticationType.SelfSigned &&
                   renderSelfSignedSection()
                }
            </div>
        );
    };

    const handleSave = () => {
        AppInsightsClient.trackUserAction(TELEMETRY_USER_ACTIONS.ADD_MODULE);
        dispatch(addModuleIdentityAction.started({
            authentication: {
                symmetricKey: authenticationType === DeviceAuthenticationType.SymmetricKey ? {
                    primaryKey: autoGenerateKeys ? null : primaryKey.value,
                    secondaryKey: autoGenerateKeys ? null : secondaryKey.value
                } : null,
                type: authenticationType.toString(),
                x509Thumbprint: authenticationType === DeviceAuthenticationType.SelfSigned ? {
                    primaryThumbprint: processThumbprint(primaryKey.thumbprint),
                    secondaryThumbprint: processThumbprint(secondaryKey.thumbprint)
                } : null,
            },
            deviceId,
            moduleId: module.id
        }));
    };

    const navigateToModuleList = () => {
        const path = pathname.replace(/\/add\/.*/, ``);
        history.push(`${path}/?${ROUTE_PARAMS.DEVICE_ID}=${encodeURIComponent(deviceId)}`);
    };

    const disableSaveButton = () => {
        if (!module.id || !validateModuleIdentityName(module.id)) { return true; }
        if (authenticationType === DeviceAuthenticationType.SymmetricKey) {
            return !isSymmetricKeyValid();
        }
        if (authenticationType === DeviceAuthenticationType.SelfSigned) {
            return !isSelfSignedThumbprintValid();
        }
        return false;
    };

    const isSymmetricKeyValid = () => {
        if (!autoGenerateKeys) {
            return primaryKey.value &&
                secondaryKey.value &&
                validateKey(primaryKey.value) &&
                validateKey(secondaryKey.value);
        }
        return true;
    };

    const isSelfSignedThumbprintValid = () => {
        return primaryKey.thumbprint &&
            secondaryKey.thumbprint &&
            validateThumbprint(primaryKey.thumbprint) &&
            validateThumbprint(secondaryKey.thumbprint);
    };

    const changeModuleIdentityName = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
        const moduleIdError = getModuleIdentityNameValidationMessage(newValue);
        setModule({
            error: moduleIdError,
            id: newValue
        });
    };

    const changeAuthenticationType = (ev?: React.FormEvent<HTMLElement | HTMLInputElement>, option?: IChoiceGroupOption) => {
        setAuthenticationType(option.key as DeviceAuthenticationType);
    };

    const changeAutoGenerateKeys = (ev?: React.FormEvent<HTMLElement | HTMLInputElement>, checked?: boolean) => {
        setautoGenerateKeys(checked);
        if (checked) {
            setPrimaryKey(initialKeyValue);
            setSecondaryKey(initialKeyValue);
        }
    };

    const changePrimaryKey = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
        const primaryKeyError = getSymmetricKeyValidationMessage(newValue);
        setPrimaryKey({
            ...primaryKey,
            error: primaryKeyError,
            value: newValue
        });
    };

    const changeSecondaryKey = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
        const secondaryKeyError = getSymmetricKeyValidationMessage(newValue);
        setSecondaryKey({
            ...secondaryKey,
            error: secondaryKeyError,
            value: newValue
        });
    };

    const changePrimaryThumbprint = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
        const primaryThumbprintError = getThumbprintValidationMessage(newValue);
        setPrimaryKey({
            ...primaryKey,
            thumbprint: newValue,
            thumbprintError: primaryThumbprintError
        });
    };

    const changeSecondaryThumbprint = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
        const secondaryThumbprintError = getThumbprintValidationMessage(newValue);
        setSecondaryKey({
            ...secondaryKey,
            thumbprint: newValue,
            thumbprintError: secondaryThumbprintError
        });
    };

    const getThumbprintValidationMessage = (value: string) => {
        return validateThumbprint(value) ? '' : ResourceKeys.moduleIdentity.validation.invalidThumbprint;
    };

    const getModuleIdentityNameValidationMessage = (value: string) => {
        return validateModuleIdentityName(value) ? '' : ResourceKeys.moduleIdentity.validation.invalidModuleIdentityName;
    };

    const getSymmetricKeyValidationMessage = (value: string): string => {
        return autoGenerateKeys ? '' : validateKey(value) ? '' : ResourceKeys.moduleIdentity.validation.invalidKey;
    };

    return (
        <>
            {showCommandBar()}
            <HeaderView
                headerText={ResourceKeys.moduleIdentity.headerText}
            />
            <div className="device-detail">
                {showModuleId()}
                {showAuthentication()}
            </div>
        </>
    );
};
