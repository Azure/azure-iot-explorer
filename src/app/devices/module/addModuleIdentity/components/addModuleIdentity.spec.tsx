/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import 'jest';
import { act } from 'react-dom/test-utils';
import { shallow, mount } from 'enzyme';
import { ChoiceGroup, CommandBar, TextField } from '@fluentui/react';
import { AddModuleIdentity } from './addModuleIdentity';
import { SynchronizationStatus } from '../../../../api/models/synchronizationStatus';
import { DeviceAuthenticationType } from '../../../../api/models/deviceAuthenticationType';
import * as AsyncSagaReducer from '../../../../shared/hooks/useAsyncSagaReducer';
import { addModuleIdentityAction } from '../actions';

const pathname = `/`;

jest.mock('react-router-dom', () => ({
    useHistory: () => ({ push: jest.fn() }),
    useLocation: () => ({ pathname, search: '' })
}));

describe('devices/components/addModuleIdentity', () => {
    beforeEach(() => {
        jest.spyOn(AsyncSagaReducer, 'useAsyncSagaReducer').mockReturnValue([{synchronizationStatus: SynchronizationStatus.initialized}, jest.fn()]);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    context('snapshot', () => {
        it('matches snapshot while loading', () => {
            expect(shallow(<AddModuleIdentity/>)).toMatchSnapshot();
        });

        it('calls add', () => {
            const addModuleIdentityActionSpy = jest.spyOn(addModuleIdentityAction, 'started');
            const wrapper = mount(<AddModuleIdentity/>);
            const commandBar = wrapper.find(CommandBar).first();
            act(() => commandBar.props().items[0].onClick(null));

            expect(addModuleIdentityActionSpy).toBeCalled();
        });

        it('renders symmetric key input field if not auto generating keys', () => {
            const wrapper = mount(<AddModuleIdentity/>);
            expect(wrapper.find(TextField).length).toEqual(1);

            // uncheck auto generate
            const autoGenerateButton = wrapper.find('.autoGenerateButton').first();
            act(() => autoGenerateButton.props().onChange?.(undefined as any));
            wrapper.update();
            expect(wrapper.find(TextField).length).toEqual(3); // tslint:disable-line:no-magic-numbers

            act(() => wrapper.find(TextField).at(1).props().onChange?.(undefined as any, 'test-key1'));
            act(() => wrapper.find(TextField).at(2).props().onChange?.(undefined as any, 'test-key2')); // tslint:disable-line:no-magic-numbers
            wrapper.update();
            const fields = wrapper.find(TextField);
            expect(fields.at(1).props().value).toEqual('test-key1');
            expect(fields.at(1).props().errorMessage).toEqual('moduleIdentity.validation.invalidKey');
            expect(fields.last().props().value).toEqual('test-key2');
            expect(fields.last().props().errorMessage).toEqual('moduleIdentity.validation.invalidKey');
        });

        it('renders selfSigned input field', () => {
            const wrapper = mount(<AddModuleIdentity/>);

            const choiceGroup = wrapper.find(ChoiceGroup).first();
            act(() => choiceGroup.props().onChange?.(undefined as any, { key: DeviceAuthenticationType.SelfSigned, text: 'text' } ));
            wrapper.update();
            expect(wrapper.find(TextField).length).toEqual(3);  // tslint:disable-line:no-magic-numbers

            act(() => wrapper.find(TextField).at(1).props().onChange?.(undefined as any, 'test-thumbprint1'));
            act(() => wrapper.find(TextField).last().props().onChange?.(undefined as any, 'test-thumbprint2'));
            wrapper.update();
            const fields = wrapper.find(TextField);
            expect(fields.at(1).props().value).toEqual('test-thumbprint1');
            expect(fields.at(1).props().errorMessage).toEqual('moduleIdentity.validation.invalidThumbprint');
            expect(fields.last().props().value).toEqual('test-thumbprint2');
            expect(fields.last().props().errorMessage).toEqual('moduleIdentity.validation.invalidThumbprint');
        });
    });
});
