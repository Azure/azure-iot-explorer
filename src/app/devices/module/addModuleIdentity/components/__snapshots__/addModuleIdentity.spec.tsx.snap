// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`devices/components/addModuleIdentity snapshot matches snapshot while loading 1`] = `
<Fragment>
  <StyledCommandBarBase
    className="command"
    items={
      Array [
        Object {
          "ariaLabel": "moduleIdentity.command.save",
          "disabled": true,
          "iconProps": Object {
            "iconName": "Save",
          },
          "key": "Save",
          "name": "moduleIdentity.command.save",
          "onClick": [Function],
        },
        Object {
          "ariaLabel": "moduleIdentity.command.cancel",
          "iconProps": Object {
            "iconName": "Cancel",
          },
          "key": "Cancel",
          "name": "moduleIdentity.command.cancel",
          "onClick": [Function],
        },
      ]
    }
  />
  <HeaderView
    headerText="moduleIdentity.headerText"
  />
  <div
    className="device-detail"
  >
    <StyledTextFieldBase
      ariaLabel="moduleIdentity.moduleId"
      description="moduleIdentity.moduleIdTooltip"
      errorMessage=""
      label="moduleIdentity.moduleId"
      onChange={[Function]}
      required={true}
      value=""
    />
    <div
      className="authentication"
    >
      <StyledChoiceGroup
        label="moduleIdentity.authenticationType.text"
        onChange={[Function]}
        options={
          Array [
            Object {
              "key": "sas",
              "text": "moduleIdentity.authenticationType.symmetricKey.type",
            },
            Object {
              "key": "selfSigned",
              "text": "moduleIdentity.authenticationType.selfSigned.type",
            },
            Object {
              "key": "certificateAuthority",
              "text": "moduleIdentity.authenticationType.ca.type",
            },
          ]
        }
        required={true}
        selectedKey="sas"
      />
      <StyledCheckboxBase
        checked={true}
        className="autoGenerateButton"
        label="moduleIdentity.authenticationType.symmetricKey.autoGenerate"
        onChange={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
